cms/lib/xblock/upstream_sync.py

```diff
@@ -0,0 +1,228 @@
+        )
+    return usage_key
+
+
+@dataclass(frozen=True)
+class UpstreamInfo:
+    """
+    Metadata about a block's relationship with an upstream.
+    """
+    usage_key: UsageKey
+    current_version: int
+    latest_version: int | None
+    sync_url: str
+    error: str | None
+
+    def sync_available(self) -> bool:
+        """
+        Should the user be prompted to sync this block with upstream?
+        """
+        return (
+            self.latest_version
+            and self.current_version < self.latest_version
+            and not self.error
+        )
+
+
+class UpstreamSyncMixin(XBlockMixin):
+    """
+    Mixed into CMS XBlocks so that they be associated & synced with uptsream (e.g. content library) blocks.
+    """
+
+    upstream = String(
+        scope=Scope.settings,
+        help=(
+            "The usage key of a block (generally within a content library) which serves as a source of upstream "
+            "updates for this block, or None if there is no such upstream. Please note: It is valid for this field "
+            "to hold a usage key for an upstream block that does not exist (or does not *yet* exist) on this instance, "
+            "particularly if this downstream block was imported from a different instance."
+        ),
+        hidden=True,
+        default=None,
+        enforce_type=True,
+    )
+    upstream_version = Integer(
+        scope=Scope.settings,
+        help=(
+            "Record of the upstream block's version number at the time this block was created from it. If "
+            "upstream_version is smaller than the upstream block's latest version, then the user will be able to sync "
+            "updates into this downstream block."
+        ),
+        hidden=True,
+        default=None,
+        enforce_type=True,
+    )
+    upstream_overridden = Set(
+        scope=Scope.settings,
+        help=(
+            "Names of the fields which have values set on the upstream block yet have been explicitly overridden "
+            "on this downstream block. Unless explicitly cleared by ther user, these overrides will persist even "
+            "when updates are synced from the upstream."
+        ),
+        hidden=True,
+        default=[],
+        enforce_type=True,
+    )
+    upstream_settings = Dict(
+        scope=Scope.settings,
+        help=(
+            "@@TODO helptext"
+        ),
+        hidden=True,
+        default={}, enforce_type=True,
+    )
+
+    def save(self, *args, **kwargs):
+        """
+        Upon save, ensure that upstream_overriden tracks all upstream-provided fields which downstream has overridden.
+
+        @@TODO use is_dirty instead of getattr for efficiency?
+        """
+        for field_name, value in self.upstream_settings.items():
+            if field_name not in self.upstream_overridden:
+                if value != getattr(self, field_name):
+                    self.upstream_overridden.append(field_name)
```

Comment:
```suggestion
                    self.upstream_overridden.add(field_name)
```
upstream_overridden is now a Set, not a List

Reply:
Punted